---
title: 'Cluster tumoral cells'
author: "Ramon Massoni-Badosa"
date: "2021/04/09"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction

From our exploratory analysis, we know that the tissue of origin (PB, lymph node or bone marrow) is a major driver of transcriptional variability. Thus, we will only work with clinical samples that come from the same niche.


# Pre-processing

## Load packages

```{r}
library(Seurat)
library(SeuratWrappers)
library(harmony)
library(tidyverse)
```


## Define parameters

```{r}
# Paths
path_to_obj <- here::here("results/R_objects/4.seurat_leukemic.rds")
path_to_save <- here::here("results/R_objects/5.seurat_list_012_019_clustered.rds")


# Colors
color_palette <- c("black", "gray", "red", "yellow", "violet", "green4",
                   "blue", "mediumorchid2", "coral2", "blueviolet",
                   "indianred4", "deepskyblue1", "dimgray", "deeppink1",
                   "green", "lightgray", "hotpink1")


# Source functions
source(here::here("bin/utils.R"))
```


## Load data

```{r}
seurat <- readRDS(path_to_obj)
```


# Subset

Let us start by visualizing our initial embedding:

```{r}
DimPlot(seurat, group.by = "tissue")
```

We will know split the cells into its donor of origin and check the tissue of origin across samples:

```{r}
seurat_list <- SplitObject(seurat, split.by = "donor_id")
purrr::walk(names(seurat_list), function(x) {
  print(x)
  print(table(seurat_list[[x]]$sample_description_FN, seurat_list[[x]]$tissue))
})
table(seurat_list[["3299"]]$sample_description_FN, seurat_list[["3299"]]$tissue)

```

The samples from donor 365 are clearly confounded by tissue. In other words, in case of observing differences between time-points, we would not know if they come from differences in niche or clinical stage. Thus, we will rule it out for now.

On the other hand, the RS sample from donor 3299 comes from PB, whilst the others come from BM. However, since we know that Ferran could assign 80 to 90% of the cells in post-treatment_3 to Richter cells, we will work with them and exclude the RS sample.

```{r}
seurat_list <- seurat_list[c("12", "19", "3299")]
seurat_list$`3299`
seurat_list$`3299` <- subset(
  seurat_list$`3299`,
  subset = sample_description_FN != "RS"
)
seurat_list$`3299`
```


# Dimensionality reduction

```{r}
seurat_list <- purrr::map(seurat_list, process_seurat, dims = 1:20)
umaps_subproject <- purrr::map(seurat_list, DimPlot, group.by = "subproject")
umaps_subproject
```

As we can see, there is a large batch effect between the two experiments we conducted BCLLATLAS_10 and BCLLATLAS_29. As we know, in BCLLATLAS_10 we obtain single-cell transcriptomes for serial clinical samples of 4 donors. However, as we observed that some points were underrepresented (had less cells), we performed scRNA-seq of specific cases (BCLLATLAS_29).

Thus, we will initially work with BCLLATLAS_10, and use BCLLATLAS_29 later as a validation.


```{r}
seurat_list <- purrr::map(seurat_list, function(seurat_obj) {
  seurat_obj <- subset(seurat_obj, subset = subproject == "BCLLATLAS_10")
  seurat_obj <- process_seurat(seurat_obj, dims = 1:20)
  seurat_obj
})
purrr::map(seurat_list, DimPlot)
```


Interestingly, we see subset of cells between Richter-like and CLL-like  cells that have a high mitochondrial content

```{r}
vars_of_interest <- c("pct_mt", "INPP5D", "MIR155HG", "CHD2")
feature_plots <- purrr::map(vars_of_interest, function(x) {
  FeaturePlot(seurat_list$`19`, x) +
    scale_colour_viridis_c(option = "inferno")
})
umap_richter <- DimPlot(seurat_list$`19`, group.by = "is_richter")
umap_richter
feature_plots
```


# Clustering

```{r}
resolutions <- c(0.1, 0.25, 0.3, 0.4, 0.5)
seurat_list <- purrr::map(
  seurat_list,
  FindNeighbors,
  reduction = "pca",
  dims = 1:20
)
seurat_list <- purrr::map(seurat_list, FindClusters, resolution = resolutions)


# Chosen resolutions
umap_clusters_12 <- DimPlot(
  seurat_list$`12`,
  group.by = "RNA_snn_res.0.25",
  cols = color_palette
)
Idents(seurat_list$`12`) <- "RNA_snn_res.0.25"
umap_clusters_19 <- DimPlot(
  seurat_list$`19`,
  group.by = "RNA_snn_res.0.25",
  cols = color_palette
)
Idents(seurat_list$`3299`) <- "RNA_snn_res.0.4"
umap_clusters_3299 <- DimPlot(
  seurat_list$`3299`,
  group.by = "RNA_snn_res.0.4",
  cols = color_palette
)

umap_clusters_12
umap_clusters_19
umap_clusters_3299
FeatureScatter(
  seurat_list$`19`,
  feature1 = "nFeature_RNA",
  feature2 = "pct_mt",
  group.by = "RNA_snn_res.0.25"
)
```


# Save

```{r}
saveRDS(seurat_list, path_to_save)
```


# Session Information

```{r}
sessionInfo()
```

