---
author: "Ramon Massoni-Badosa"
date: "2020/04/10"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
title <- paste("Hashtag Demultiplexing: ", library_name, sep = "")
```

---
title: "`r title`"
---

# Introduction

We had problems to demultiplex the cells from the patient ICGC_3299. This is because a specific  sample (ICGC_3299_19_2768B) does not contain any counts for its corresponding hashtag oligonucleotide (HTO). This can be due to the following challenges:

- The labeling of this sample didn't work because the antibody was degraded or did not function properly. In this case our dataset can have a considerable number of cells of this sample, which can challenge interpretation if we cannot assign them to their original condition properly.
- There were too few cells to tag, so our dataset has no cells which belong to this condition. In this case, we can safely proceed with our analysis.

Here, we will exclude the HTO reads of this sample and demultiplex the rest to shed light into which of these two scenarios we are in.


## Load packages

```{r}
library(Seurat)
library(scales)
library(tidyverse)
```


## Parameters

```{r}
# Paths
path_to_data <- str_c(
  "1-cellranger_mapping/projects/BCLLATLAS_10",
  "/jobs/",
  gem_id,
  "/",
  gem_id,
  "/outs/filtered_feature_bc_matrix/"
)
path_to_data <- here::here(path_to_data)
print(path_to_data)
path_to_scrublet <- here::here(str_c(
  "results/tables/scrublet/scrublet_doublet_prediction-BCLLATLAS_10-",
  gem_id,
  ".csv"
))


# Functions
source(here::here("bin/utils.R"))
```


## Load data

```{r}
matrices <- Read10X(data.dir = path_to_data)
```


# Demultiplex

To demultiplex, we will follow the ["Demultiplexing with hashtag oligos (HTOs)" pipeline from Seurat](https://satijalab.org/seurat/v3.0/hashing_vignette.html):

```{r}
# Setup Seurat object
seurat <- CreateSeuratObject(counts = matrices$`Gene Expression`)


# Add HTO as an independent assay, eliminating ICGC_3299_19_2768B
selected_rows <- rownames(matrices$`Antibody Capture`) != "ICGC_3299_19_2768B"
matrices$`Antibody Capture` <- matrices$`Antibody Capture`[selected_rows, ] 
seurat[["HTO"]] <- CreateAssayObject(counts = matrices$`Antibody Capture`)
seurat <- NormalizeData(seurat, assay = "HTO", normalization.method = "CLR")


# Demultiplex
seurat <- HTODemux(seurat, assay = "HTO", positive.quantile = 0.99)
new_barcodes <- str_c(gem_id, colnames(seurat), sep = "_")
seurat <- RenameCells(seurat, new.names = new_barcodes)
```


# Visualization

We can visualize the results as ridge plots and heatmaps:

## Ridge plots

```{r fig.height=10, fig.width=14}
table(seurat$HTO_classification.global)
Idents(seurat) <- "HTO_maxID"
RidgePlot(
  object = seurat, 
  assay = "HTO", 
  features = rownames(seurat[["HTO"]]), 
  ncol = 3
)
```

The previous step classified each cell in one or more batch. We can see the levels of each hashtag across the different batches that the algorithm classified. Ideally, we should observe a large signal-to-noise ratio. That is, the HTO is only highly expressed in the batch that it identifies and 0 in the others.


## Heatmaps

```{r}
HTOHeatmap(seurat, assay = "HTO", ncells = 5000)
```

In this heatmap, each column represents a 10X barcode, and each row the expression of a batch-specific HTO. We can visualize singlets, doublets and negative cells (ie not labelled by any HTO). We strive for a similar number of cells across batches, as this increases our ability to identify doublets.


## Number of cells

Let us visualize the numbers of singlets/doublets/negative:

```{r}
# Absolute frequencies
hto_levels <- c("Singlet", "Doublet", "Negative")
seurat_df <- seurat@meta.data %>%
  group_by(HTO_classification.global) %>% 
  summarise(count = n()) %>%
  mutate(percentage = (count / sum(count)) * 100) %>%
  mutate(HTO_classification.global = factor(
    HTO_classification.global, 
    levels = hto_levels))
seurat_count_gg <- seurat_df %>% 
  ggplot(aes(HTO_classification.global, count, fill = HTO_classification.global)) +
  geom_col() +
  geom_text(aes(label = count), 
            position = position_dodge(width = 0.9), vjust = -0.25) +
  labs(title = gem_id, x = "", y = "number of cells") +
  theme_bw() +
  scale_fill_brewer(palette = "Dark2") +
  theme(axis.text.x = element_text(size = 11), 
        plot.title = element_text(hjust = 0.5),
        legend.position = "none")
seurat_count_gg


# Percentages
seurat_pct_gg <- seurat_df %>% 
  ggplot(aes(x = "", y = percentage, fill = HTO_classification.global)) +
    geom_bar(width = 1, stat = "identity") +
    geom_text(
      aes(label = percent(percentage / 100)),
      position = position_stack(vjust = 0.5),
      size = 4
    ) +
    scale_fill_brewer(palette = "Dark2") +
    coord_polar("y", start = 0) +
    theme_minimal() +
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      panel.border = element_blank(),
      panel.grid = element_blank(),
      axis.ticks = element_blank(),
      axis.text.x = element_blank(),
      legend.title = element_blank(),
      legend.text = element_text(size = 11)
    )
seurat_pct_gg
```



# Scrublet accuracy

With cell hashing we cannot detect intra-batch doublets. To assess it, we ran [scrublet](https://www.cell.com/cell-systems/pdfExtended/S2405-4712(18)30474-5) in the previous step. Given that hashing gives us ground-truth about the doublets in the dataset, we can assess the accuracy of scrublet:

```{r}
# Load scrublet data
scrublet_df <- read_csv(path_to_scrublet)


# Add scrublet doublet score to metadata
scrublet_df$barcodes <- str_c(gem_id, scrublet_df$barcodes, sep = "_")
if (all(scrublet_df$barcodes == colnames(seurat))) {
  warning("barcodes are equal")
  seurat$scrublet_doublet_scores <- scrublet_df$scrublet_doublet_scores
  seurat$scrublet_doublet_scores_scaled <- scale(
    scrublet_df$scrublet_doublet_scores,
    center = TRUE,
    scale = TRUE
  )
  seurat$scrublet_predicted_doublet <- scrublet_df$scrublet_predicted_doublet
  
} else{
  warning("barcodes are not equal")
}


# Plot doublet score distribution
score_gg <- seurat@meta.data %>%
    mutate(HTO_classification.global = factor(
      seurat$HTO_classification.global,
      levels = c("Singlet", "Doublet", "Negative")
    )) %>% 
    ggplot(aes(
      HTO_classification.global,
      scrublet_doublet_scores,
      color = HTO_classification.global
    )) +
      labs(x = "", y = "scrublet doublet score")
score_gg <- customized_boxplot(score_gg) 
score_gg
```


# Quantify the signal-to-noise ratio (SNR)

We quantify the SNR in the cell hashing experiment by dividing, for each cell, the first and second highest HTO normalized counts. In the case of singlets we should expect a clear deviation from 1:

```{r}
seurat$hashing_snr <- purrr::map_dbl(colnames(seurat), function(x) {
  signal <- seurat[["HTO"]]@data[seurat$HTO_maxID[x], x]
  noise <- seurat[["HTO"]]@data[seurat$HTO_secondID[x], x]
  snr <- (signal + 0.1) / (noise + 0.1)
  snr
})
snr_gg <- seurat@meta.data %>%
    mutate(HTO_classification.global = factor(
      seurat$HTO_classification.global,
      levels = c("Singlet", "Doublet", "Negative")
    )) %>% 
    ggplot(aes(
      HTO_classification.global,
      hashing_snr,
      color = HTO_classification.global
    )) +
      labs(x = "", y = "signal-to-noise")
snr_gg <- customized_boxplot(snr_gg)
snr_gg
```


# Save demultiplexed Seurat objects

```{r}
saveRDS(seurat, save_object_path)
```


# Session Info

```{r}
sessionInfo()
```
